### PR TITLE
Release note for 1.15.0

### PR DESCRIPTION
**Pull request by trustin**
- Added the release note for 1.15.0
- Fixed a bug where `:site:apiIndex` handles a method with varargs
  incorrectly
- Fixed a bug where the release note auto-generate script doesn't pad
  month and day with `0`
  - Also made sure the date in frontmatter is always parsed strictly to
    avoid any ambiguity
- Removed a reference to a non-existent CSS style in
  `site/src/layouts/header.tsx`
- Increased the bottom margin of an alert box when it comes as the last
  item in a list for aesthetics' sake
- Added a section about how to auto-generate a release note to
  `site/README.md`
- Release notes and news pages now show a 'tip' that links to the latest
  release note or news item, to let the users know they are reading an
  old release note or news item.

### CODE REVIEW

**Comment by trustin** [comment]
[Preview (PDF)](https://github.com/line/armeria/files/8343078/1.15.0.release.notes.pdf)

File: site/src/pages/release-notes/1.15.0.mdx [hunk 1]
```diff
@@ -0,0 +1,351 @@
+---
+date: 2022-03-24
+---
+
+## 🌸 Highlights
+
+- gRPC health check protocol
+- Multipart file uploads in annotated services.
```
Comment (trustin) [inline site/src/pages/release-notes/1.15.0.mdx:8]:
```suggestion
- Multipart file uploads in annotated services
```

File: site/README.md [hunk 2]
```diff
@@ -92,3 +92,9 @@ how much space using `source-map-explorer`.
 2. A web browser will show a tree map.
    See [here](https://github.com/danvk/source-map-explorer#readme) to
    learn more about how to interpret the report.
+
+### Generating release notes
+
+```console
+$ npm run release-note <version>
```
Comment (ikhoon) [inline site/README.md:99]: 👍

File: site/src/pages/release-notes/1.15.0.mdx [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+    - The number of days before certificate expires, or `-1` if expired.
+- You can now specify a `File` or `Path` parameter in your annotated service method to handle multipart file
+  uploads. #3578 #4061
+  ```java
+  Server
+    .builder()
+    .annotatedService(new Object() {
+      @Post("/upload")
+      @Consumes("multipart/form-data")
+      String upload(@Param File file) {
```
Comment (ikhoon) [inline site/src/pages/release-notes/1.15.0.mdx:29]:
```suggestion
      public String upload(@Param File file) {
```

File: site/src/pages/release-notes/1.15.0.mdx [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+        return "Successfully uploaded a file to " + file;
+      }
+    })
+    .build();
+  ```
+- You can now use the `||` (logical OR) operator in parameter and header predicate expressions. #4116 #4138
+  ```java
+  // "Hello!" for /greet or /greet?casual=false
+  // "Hi!" for /greet?casual or /greet?casual=true
```
Comment (ikhoon) [inline site/src/pages/release-notes/1.15.0.mdx:38]: Follow the order of routes?
```suggestion
  // "Hi!" for /greet?casual or /greet?casual=true
  // "Hello!" for /greet or /greet?casual=false
```

File: site/src/pages/release-notes/1.15.0.mdx [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+      .matchesParam("casual || casual=true") // 👈👈👈
+      .build((ctx, req) -> HttpResponse.of("Hi!"))
+    .route()
+      .get("/greet")
+      .matchesParam("!casual || casual!=true") // 👈👈👈
+      .build(ctx, req) -> HttpResponse.of("Hello!"))  
+  ```
+  <Tip>
+  
+  You can use the same syntax in [annotated services](../docs/server-annotated-service)
```
Comment (ikhoon) [inline site/src/pages/release-notes/1.15.0.mdx:52]:
```suggestion
  You can use the same syntax in [annotated services](/docs/server-annotated-service)
```

File: site/src/pages/release-notes/1.15.0.mdx [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+  </Tip>
+- You can now enable [gRPC health check protocol](https://github.com/grpc/grpc/blob/master/src/proto/grpc/health/v1/health.proto)
+  with <type://GrpcHealthCheckService> #3146 #3963
+  ```java
+  Server
+    .builder()
+    .service(
+      GrpcService
+        .builder()
+        .addService(GrpcHealthCheckService.of()) // 👈👈👈
```
Comment (ikhoon) [inline site/src/pages/release-notes/1.15.0.mdx:64]: Should it be better to show `enableHealthCheckService(boolean)` as a sensible default option?

File: site/src/pages/release-notes/1.15.0.mdx [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+    .build();
+  ```
+- Armeria now has the global shared DNS cache by default for higher hit ratio and less DNS traffic.
+  You can also build and share a custom DNS cache across multiple resolvers. #2940 #4133
+  ```java
+  DnsCache dnsCache = DnsCache
+    .builder()
+    .ttl(60, 3600)
+    .negativeTtl(600)
+    .build();
```
Comment (ikhoon) [inline site/src/pages/release-notes/1.15.0.mdx:98]: nit: For the consistent code format with the followings?
```suggestion
  DnsCache dnsCache = 
    DnsCache
      .builder()
      .ttl(60, 3600)
      .negativeTtl(600)
      .build();
```

File: site/src/pages/release-notes/1.15.0.mdx [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+    .newDecorator()    
+  ```
+- You can now specify a customizer that customizes a <type://ClientRequestContext> when building a client. #4075 #4082
+  ```java
+  WebClient client =
+    WebClient
+      .builder()
+      .contextCustomizer(ctx -> { // 👈👈👈
+        String userId = MyUserContext.current().getId();
+          ctx.setAttr(USER_ID, userId);
```
Comment (ikhoon) [inline site/src/pages/release-notes/1.15.0.mdx:184]:
```suggestion
        String userId = MyUserContext.current().getId();
        ctx.setAttr(USER_ID, userId);
```

File: site/src/pages/release-notes/1.15.0.mdx [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+  Tracing requestContextTracing =
+    Tracing
+      .newBuilder()
+      .currentTraceContext(RequestContextCurrentTraceContext.ofDefault())
+      .build();
+
+  WebClient
+    .builder()
+    .contextCustomizer(TraceContextPropagation.inject(() -> {
+      // Propagate Reactor's TracingContext to Armeria's TractContext.
```
Comment (ikhoon) [inline site/src/pages/release-notes/1.15.0.mdx:200]:
```suggestion
      // Propagate Reactor's TraceContext to Armeria's TraceContext.
```

**Review by ikhoon — COMMENTED** [review]
Wow, huge changelogs!

File: site/src/pages/release-notes/1.15.0.mdx [hunk 10]
```diff
@@ -0,0 +1,351 @@
...
+- Prometheus 0.14.1 → 0.15.0
+- Reactor 3.4.14 → 3.4.16
+  - Reactor Kotlin 1.1.5 → 1.1.6
+- RxJava 3.1.3 → 3.1.4
+- Sangria 2.1.6 → 3.0.0
+- ScalaPB 0.11.8 → 0.11.10
+- scala-collection-compat 2.6.0 → 2.7.0
+- SLF4J 1.7.34 → 1.7.36
+- Spring 5.3.15 → 5.3.17
+- Spring Boot 2.6.3 → 2.6.4
```
Comment (ikhoon) [inline site/src/pages/release-notes/1.15.0.mdx:326]: A new Spring Boot version is released.
```suggestion
- Spring Boot 2.6.3 → 2.6.5
```

**Review by jrhee17 — APPROVED** [review]
Happy with the changes as long as @ikhoon 's comments are addressed 👍  Thanks @trustin ! 🙇 👍 🙇

File: site/src/pages/release-notes/1.15.0.mdx (hunk 6 above)
Comment (trustin) [reply site/src/pages/release-notes/1.15.0.mdx]: Good point. Updated

**Review by ikhoon — APPROVED** [review]
Thanks!

**Pull request merged**